// @react-pdf-viewer/scroll-mode/lib/cjs/scroll-mode.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var core = require('@react-pdf-viewer/core');

function _interopNamespace(e) {
    if (e && e.__esModule) return e;
    var n = Object.create(null);
    if (e) {
        Object.keys(e).forEach(function (k) {
            if (k !== 'default') {
                var d = Object.getOwnPropertyDescriptor(e, k);
                Object.defineProperty(n, k, d.get ? d : {
                    enumerable: true,
                    get: function () { return e[k]; }
                });
            }
        });
    }
    n["default"] = e;
    return Object.freeze(n);
}

var React__namespace = /*#__PURE__*/_interopNamespace(React);

exports.ScrollMode = void 0;
(function (ScrollMode) {
    ScrollMode["Horizontal"] = "Horizontal";
    ScrollMode["Vertical"] = "Vertical";
    ScrollMode["Wrapped"] = "Wrapped";
})(exports.ScrollMode || (exports.ScrollMode = {}));

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */

var __assign = function() {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p)) t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

var HorizontalScrollingIcon = function () { return (React__namespace.createElement(core.Icon, { size: 16 },
    React__namespace.createElement("path", { d: "M6.5,21.5c0,0.552-0.448,1-1,1h-4c-0.552,0-1-0.448-1-1v-20c0-0.552,0.448-1,1-1h4c0.552,0,1,0.448,1,1V21.5z\n            M14.5,21.5c0,0.552-0.448,1-1,1h-4c-0.552,0-1-0.448-1-1v-20c0-0.552,0.448-1,1-1h4c0.552,0,1,0.448,1,1V21.5z\n            M22.5,21.5 c0,0.552-0.448,1-1,1h-4c-0.552,0-1-0.448-1-1v-20c0-0.552,0.448-1,1-1h4c0.552,0,1,0.448,1,1V21.5z" }))); };

var VerticalScrollingIcon = function () { return (React__namespace.createElement(core.Icon, { size: 16 },
    React__namespace.createElement("path", { d: "M23.5,5.5c0,0.552-0.448,1-1,1h-21c-0.552,0-1-0.448-1-1v-3c0-0.552,0.448-1,1-1h21c0.552,0,1,0.448,1,1V5.5z\n            M23.5,13.5c0,0.552-0.448,1-1,1h-21c-0.552,0-1-0.448-1-1v-3c0-0.552,0.448-1,1-1h21c0.552,0,1,0.448,1,1V13.5z\n            M23.5,21.5 c0,0.552-0.448,1-1,1h-21c-0.552,0-1-0.448-1-1v-3c0-0.552,0.448-1,1-1h21c0.552,0,1,0.448,1,1V21.5z" }))); };

var WrappedScrollingIcon = function () { return (React__namespace.createElement(core.Icon, { size: 16 },
    React__namespace.createElement("path", { d: "M10.5,9.5c0,0.552-0.448,1-1,1h-8c-0.552,0-1-0.448-1-1v-8c0-0.552,0.448-1,1-1h8c0.552,0,1,0.448,1,1V9.5z\n            M23.5,9.5c0,0.552-0.448,1-1,1h-8c-0.552,0-1-0.448-1-1v-8c0-0.552,0.448-1,1-1h8c0.552,0,1,0.448,1,1V9.5z\n            M10.5,22.5 c0,0.552-0.448,1-1,1h-8c-0.552,0-1-0.448-1-1v-8c0-0.552,0.448-1,1-1h8c0.552,0,1,0.448,1,1V22.5z\n            M23.5,22.5c0,0.552-0.448,1-1,1 h-8c-0.552,0-1-0.448-1-1v-8c0-0.552,0.448-1,1-1h8c0.552,0,1,0.448,1,1V22.5z" }))); };

var SwitchScrollModeDecorator = function (_a) {
    var children = _a.children, mode = _a.mode, onClick = _a.onClick;
    var l10n = React__namespace.useContext(core.LocalizationContext).l10n;
    var label = '';
    var icon = React__namespace.createElement(VerticalScrollingIcon, null);
    switch (mode) {
        case exports.ScrollMode.Horizontal:
            label = (l10n && l10n.scrollMode ? l10n.scrollMode.horizontalScrolling : 'Horizontal scrolling');
            icon = React__namespace.createElement(HorizontalScrollingIcon, null);
            break;
        case exports.ScrollMode.Wrapped:
            label = (l10n && l10n.scrollMode ? l10n.scrollMode.wrappedScrolling : 'Wrapped scrolling');
            icon = React__namespace.createElement(WrappedScrollingIcon, null);
            break;
        case exports.ScrollMode.Vertical:
        default:
            label = (l10n && l10n.scrollMode ? l10n.scrollMode.verticalScrolling : 'Vertical scrolling');
            icon = React__namespace.createElement(VerticalScrollingIcon, null);
            break;
    }
    return children({ icon: icon, label: label, onClick: onClick });
};

var TOOLTIP_OFFSET = { left: 0, top: 8 };
var SwitchScrollModeButton = function (_a) {
    var isSelected = _a.isSelected, mode = _a.mode, onClick = _a.onClick;
    return (React__namespace.createElement(SwitchScrollModeDecorator, { mode: mode, onClick: onClick }, function (props) { return (React__namespace.createElement(core.Tooltip, { ariaControlsSuffix: "scroll-mode-switch", position: core.Position.BottomCenter, target: React__namespace.createElement(core.MinimalButton, { ariaLabel: props.label, isSelected: isSelected, onClick: props.onClick }, props.icon), content: function () { return props.label; }, offset: TOOLTIP_OFFSET })); }));
};

var useScrollMode = function (store) {
    var _a = React__namespace.useState(store.get('scrollMode') || exports.ScrollMode.Vertical), scrollMode = _a[0], setScrollMode = _a[1];
    var switchTo = function (newScrollMode) {
        setScrollMode(newScrollMode);
        var getPagesContainer = store.get('getPagesContainer');
        if (!getPagesContainer) {
            return;
        }
        var pagesEle = getPagesContainer();
        if (!pagesEle) {
            return;
        }
        switch (newScrollMode) {
            case exports.ScrollMode.Vertical:
                pagesEle.classList.add('rpv-scroll-mode__vertical');
                pagesEle.classList.remove('rpv-scroll-mode__horizontal');
                pagesEle.classList.remove('rpv-scroll-mode__wrapped');
                break;
            case exports.ScrollMode.Horizontal:
                pagesEle.classList.add('rpv-scroll-mode__horizontal');
                pagesEle.classList.remove('rpv-scroll-mode__vertical');
                pagesEle.classList.remove('rpv-scroll-mode__wrapped');
                break;
            case exports.ScrollMode.Wrapped:
                pagesEle.classList.add('rpv-scroll-mode__wrapped');
                pagesEle.classList.remove('rpv-scroll-mode__vertical');
                pagesEle.classList.remove('rpv-scroll-mode__horizontal');
                break;
        }
    };
    return { scrollMode: scrollMode, switchTo: switchTo };
};

var SwitchScrollMode = function (_a) {
    var children = _a.children, mode = _a.mode, store = _a.store;
    var _b = useScrollMode(store), scrollMode = _b.scrollMode, switchTo = _b.switchTo;
    var onClick = function () {
        switchTo(mode);
        store.update('scrollMode', mode);
    };
    var isSelected = scrollMode === mode;
    var defaultChildren = function (props) { return (React__namespace.createElement(SwitchScrollModeButton, { isSelected: isSelected, mode: props.mode, onClick: props.onClick })); };
    var render = children || defaultChildren;
    return render({
        isSelected: isSelected,
        mode: mode,
        onClick: onClick,
    });
};

var SwitchScrollModeMenuItem = function (_a) {
    var isSelected = _a.isSelected, mode = _a.mode, onClick = _a.onClick;
    return (React__namespace.createElement(SwitchScrollModeDecorator, { mode: mode, onClick: onClick }, function (props) { return (React__namespace.createElement(core.MenuItem, { checked: isSelected, icon: props.icon, onClick: props.onClick }, props.label)); }));
};

var Tracker = function (_a) {
    var store = _a.store;
    var direction = React__namespace.useContext(core.ThemeContext).direction;
    var isRtl = direction === core.TextDirection.RightToLeft;
    var _b = useScrollMode(store), scrollMode = _b.scrollMode, switchTo = _b.switchTo;
    var handlePagesContainer = function (getPagesContainer) {
        var pagesEle = getPagesContainer();
        if (!pagesEle) {
            return;
        }
        isRtl ? pagesEle.classList.add('rpv-scroll-mode--rtl') : pagesEle.classList.remove('rpv-scroll-mode--rtl');
        var scrollMode = store.get('scrollMode') || exports.ScrollMode.Vertical;
        if (scrollMode !== exports.ScrollMode.Vertical) {
            switchTo(scrollMode);
        }
    };
    var handleScrollModeChanged = function (newScrollMode) {
        if (newScrollMode !== scrollMode) {
            switchTo(newScrollMode);
        }
    };
    React__namespace.useEffect(function () {
        store.subscribe('getPagesContainer', handlePagesContainer);
        store.subscribe('scrollMode', handleScrollModeChanged);
        return function () {
            store.unsubscribe('getPagesContainer', handlePagesContainer);
            store.unsubscribe('scrollMode', handleScrollModeChanged);
        };
    }, []);
    return React__namespace.createElement(React__namespace.Fragment, null);
};

var scrollModePlugin = function (props) {
    var store = React__namespace.useMemo(function () {
        return core.createStore({
            scrollMode: props && props.scrollMode ? props.scrollMode : exports.ScrollMode.Vertical,
        });
    }, []);
    var SwitchScrollModeDecorator = function (props) { return React__namespace.createElement(SwitchScrollMode, __assign({}, props, { store: store })); };
    var SwitchScrollModeButtonDecorator = function (props) { return (React__namespace.createElement(SwitchScrollModeDecorator, { mode: props.mode }, function (p) { return (React__namespace.createElement(SwitchScrollModeButton, { isSelected: p.isSelected, mode: p.mode, onClick: function () {
            p.onClick();
        } })); })); };
    var SwitchScrollModeMenuItemDecorator = function (props) { return (React__namespace.createElement(SwitchScrollModeDecorator, { mode: props.mode }, function (p) { return (React__namespace.createElement(SwitchScrollModeMenuItem, { isSelected: p.isSelected, mode: p.mode, onClick: function () {
            p.onClick();
            props.onClick();
        } })); })); };
    var renderViewer = function (props) {
        var currentSlot = props.slot;
        if (currentSlot && currentSlot.children) {
            currentSlot.children = (React__namespace.createElement(React__namespace.Fragment, null,
                React__namespace.createElement(Tracker, { store: store }),
                currentSlot.children));
        }
        return currentSlot;
    };
    return {
        install: function (pluginFunctions) {
            store.update('getPagesContainer', pluginFunctions.getPagesContainer);
        },
        renderViewer: renderViewer,
        switchScrollMode: function (mode) {
            store.update('scrollMode', mode);
        },
        SwitchScrollMode: SwitchScrollModeDecorator,
        SwitchScrollModeButton: SwitchScrollModeButtonDecorator,
        SwitchScrollModeMenuItem: SwitchScrollModeMenuItemDecorator,
    };
};

exports.HorizontalScrollingIcon = HorizontalScrollingIcon;
exports.VerticalScrollingIcon = VerticalScrollingIcon;
exports.WrappedScrollingIcon = WrappedScrollingIcon;
exports.scrollModePlugin = scrollModePlugin;
